shiny/render/_data_frame_utils/_selection.py

```diff
@@ -0,0 +1,450 @@
+from __future__ import annotations
+
+# TODO-barret; Docs
+# TODO-barret; Add examples!
+import warnings
+from typing import Literal, Sequence, Union
+
+from ..._deprecated import warn_deprecated
+from ..._typing_extensions import NotRequired, TypedDict, TypeGuard
+
+# RowSelectionMode = Literal["none", "row", "rows"]
+RowSelectionMode = Literal["row", "rows"]
+# ColSelectionMode = Literal["col", "cols"]
+# RegionSelectionMode = Literal["region"]
+# CellSelectionMode = Literal["cell"]
+
+SelectionMode = Union[
+    RowSelectionMode,
+    # ColumnSelectionMode,
+    # RegionSelectionMode,
+    # CellSelectionMode,
+    Literal["none"],
+]
+
+LocationRange = tuple[int, int]
+
+
+# class SelectionLocation_(TypedDict):
+#     rows: LocationRange | Literal["all"]
+#     cols: LocationRange | Literal["all"]
+
+row_idx = 0
+col_idx = 9
+min_row = 0
+max_row = 3
+min_col = 0
+max_col = 3
+
+
+# Select single cell
+x = {"rows": row_idx, "cols": col_idx}
+x = {"rows": [row_idx], "cols": [col_idx]}
+# select single region
+x = {"rows": [min_row, max_row], "cols": [max_col, min_col]}
+# No multi-region selection!
+
+# select single row; all columns
+x = {"rows": row_idx}
+x = {"rows": row_idx, "cols": None}
+x = {"rows": [row_idx]}
+x = {"rows": [row_idx], "cols": None}
+# select multiple rows; all columns
+x = {"rows": [row_idx, row_idx + 3, row_idx + 5]}
+x = {"rows": [row_idx, row_idx + 3, row_idx + 5], "cols": None}
+
+# (same for cols)
+# select single col; all rows
+x = {"cols": col_idx}
+x = {"cols": col_idx, "rows": None}
+x = {"cols": [col_idx]}
+x = {"cols": [col_idx], "rows": None}
+# select multiple cols; all rows
+x = {"cols": [col_idx, col_idx + 3, col_idx + 5]}
+x = {"cols": [col_idx, col_idx + 3, col_idx + 5], "rows": None}
+
```

Comment:
Seems like there's a lot of debugging/testing code in this file which can be removed.

Reply:
Commented code has been reduced